### ISSUE TITLE
Model- and data-specific unit tests

### ISSUE THREAD

**Issue by addisonklinke** [PyTorchLightning/pytorch-lightning#5485]
## 🚀 Feature

Trainer option to run callbacks for model- and data-specific unit testing

### Motivation

The [docs](https://pytorch-lightning.readthedocs.io/en/stable/new-project.html#debugging) mention a `fast_dev_run` option for initializing a trainer. Correct me if I'm wrong, but my understanding is this simply hits every line of code in the train/val loops to ensure there won't be any crashes on a full-scale training. This seems to fall a bit short of proper unit testing that would validate several assertions in addition to confirming there are no crashes.

Unit testing for machine learning models appears to be an ongoing and relatively immature development in the community. However, there are a few useful articles discussing its use and importance 

1. PyTorch Lightning maintainer Jeremy Jordan wrote a helpful [article](https://www.jeremyjordan.me/testing-ml/) that suggests several tests
2. This Medium [article](https://medium.com/@keeper6928/mltest-automatically-test-neural-network-models-in-one-function-call-eb6f1fa5019d) which was originally written for TensorFlow, but already has a PyTorch port [here](https://github.com/suriyadeepan/torchtest)
3. Additional unit tests ideas are brainstormed in this [article](https://krokotsch.eu/cleancode/2020/08/11/Unit-Tests-for-Deep-Learning.html#what-should-i-test)

### Pitch

1. Define a set of model- and data-specific unit tests using the above articles as a starting point
2. Separate the information needed to run the tests into two categories
    
    * Generic: consistent across all models/dataset, can be deduced within the Lightning framework automatically
    * Specific: depends on the model and/or dataset - practitioner is expected to provide this

3. Determine an appropriate set of `pl.LightningModule` callbacks to allow users to provide the specific inputs discussed above
4. Use these callbacks in `pl.Trainer(do_unit_tests=True)` to execute the tests

    * I would suggest tests are on by default
    * If the callbacks have not been implemented, skip the tests and provide a logging message to the user to encourage them to define the callbacks

### Alternatives

1. Use the existing `fast_dev_run` - I explain above why this is different from proper unit tests
2. Allow users to integrate with a 3rd party test library (like the link above for `torchtest`) - I think it would be preferable to have this functionality built-into the Lightning API rather than relying on other imports. Doing so would make it easier to automate the generic inputs discussed in the pitch 
3. Expect users to write their own test suites - PyTorch Lightning strives to minimize boilerplate code and this seems like an excellent candidate for that!

**Comment by addisonklinke** [comment]
@Borda Just checking in to see if the team has given this any thought. I could likely provide some hours to write the required code, but wanted to make sure we're all on the same page first!

**Comment by addisonklinke** [comment]
@williamFalcon @awaelchli Anyone have opinions on whether this request/approach would be useful? I think it could be an excellent feature to help distinguish Lightning from other high-level ML libraries

**Comment by williamFalcon** [comment]
I like this a lot. Let's definitely take a look at integrating this.

@edenafek @tchaton

**Comment by addisonklinke** [comment]
Excellent, glad to hear you're on board. Below is a table with my detailed goals for the integration. For completeness and consistency with the references, I listed all trainer related tests even though I'd bet the majority are already implemented in Lightning's existing set of internal unit tests 

A few notes

- Reference links (under the table) provide more detailed discussion for the reasoning behind each test
- Attribute names
  - Ones that already exist are marked with `*`
  - Some assertions won't require user-defined callbacks in order to execute the unit test, those are listed as N/A     
  - My suggestions might be a bit more verbose than what's actually necessary for the final code
- All data-related tests should run independently (with shared, re-useable code) on all splits of the dataset (train, val, test)
- A generic tensor [validator](https://github.com/addisonklinke/pytorch-architectures/blob/master/torcharch/types.py#L66) could be re-used to handle commonalities between the model + dataset assertions

Let me know what all the team would consider in scope

| Item | Component | Assertion | Attribute | Reference |
| :--- | :-------- | :-------- | :-------- | :-------- |
| 1  | Model | Input matches expected shape | `lm.example_input_array` * | 
| 2  | Model | Input has acceptable numeric range/dtype | `lm.example_input_spec` | 
| 3  | Model | Output matches expected shape | `lm.example_output_array -> Union[torch.Tensor, list, dict] ` | 1, 2
| 4  | Model | Output has acceptable numeric range/dtype | `lm.example_output_spec` | 1, 3
| 5  | Model | Parameters are updatable | `lm.parameters()` * | 2, 3
| 6  | Model | Buffers do NOT update | `lm.static_parameters()` | 3
| 7  | Model | Custom loss function works | `lm.example_loss_?` | 2
| 8  | Dataset | Number of elements (i.e. image + label, etc) per batch | `dm.batch_elements -> list[BatchElementSpec]` |
| 9  | Dataset | Batch element shapes | `dm.batch_elements` | 2
| 10  | Dataset | Batch element numeric range/dtype | `dm.batch_elements` | 2
| 11 | Dataset | Repeated load + augment cycles produce same result | N/A | 2
| 12 | Dataset | No label leakage between subsets | `dm.element_uuid_map[idx] -> str` | 1
| 13 | Trained model | Invariance to specified data perturbations | `lm.data_perturbations -> list[PerturbationSpec]` | 1
| 14 | Trained model | Directional response expectation to specified data perturbations | `lm.data_perturbations -> list[PerturbationSpec]` | 1
| 15 | Trainer | Loss decreases after training step | N/A | 1
| 16 | Trainer | Moving tensors between devices (CPU <--> GPU) | N/A | 2
| 17 | Trainer | Ensure random seed maintains control over all components | N/A | 2
| 18 | Dataset/trainer | Single + multi-process data loaders | N/A | 2
| 19 | Dataset/model | Indepdence of batch samples | N/A | 2

References:
1. [Jeremy Jordan](https://www.jeremyjordan.me/testing-ml/#how-do-you-write-model-tests)
2. [Don't Repeat Yourself](https://krokotsch.eu/cleancode/2020/08/11/Unit-Tests-for-Deep-Learning.html#what-should-i-test)
3. [`mltest/torchtest`](https://medium.com/@keeper6928/mltest-automatically-test-neural-network-models-in-one-function-call-eb6f1fa5019d)

**Comment by addisonklinke** [comment]
Just curious how this is looking for the 1.4 milestone? Does the team need any more discussion from me to help finalize the scope?

**Comment by ananthsub** [comment]
I'd like to better understand the central assumptions in the issue: 

Is this more about what Lightning can do in designing its interfaces to be more testable? Or to provide out of the box testing utilities for users?

I agree with the former: Lightning as a framework should be easily testable (unit/integration/etc) in the context of user applications. This is also needed from the core library POV to ensure things work as expected. 

However, offering "out of the box" testing utilities is much more difficult because of how application-dependent these could instances can be, especially with the generality of these interfaces. 

> Determine an appropriate set of pl.LightningModule callbacks to allow users to provide the specific inputs discussed above

What determines this? This again feels closer to per-Lightning Module utilities instead of generic ones offered by the framework. 

> Input has acceptable numeric range/dtype

For instance, hat's an example of this?

> No label leakage between subsets

Why not test this at the Dataset level instead of the datamodule? The DataModule is a wrapper around this for the dataloader

> Use these callbacks in pl.Trainer(do_unit_tests=True) to execute the tests

I do not like having a separate codepath for unit test execution. A separate flag could easily introduce different behavior for unit testing vs actual training, which means tests can pass and the real run still breaks.

Additionally, when you refer to callbacks, do you mean Lightning callbacks? I don't think these should be required for testing. Rather, we should encourage unit-testing of components as much as possible to remove dependencies. Otherwise we're running mini end to end tests using the Trainer + callbacks for each data/model assertion.

**Comment by carmocca** [comment]
The path forward here would be to improve the flexibility and testability of Lightning and its components. Adding a constrained set of internal "tests" would be limiting as it would hold strong assumptions of the model/data/component internal behaviour and their interactions.

I think most of the "test cases" described can be created by the users externally to the Lightning codebase.

If you are still interested in this, I would suggest trying to write them to see why, where and how Lightning can be reshaped so its flexibility and testability improve. Additionally, you could add test cases to https://github.com/PyTorchLightning/ecosystem-ci to ensure compatibility is kept for your specific model/data/component.

Since this conversation is stale, I will close the issue. But feel free to continue the discussion.
